File: app/android/app/src/main/java/com/mrnobody/agent/planner/IntentRouter.java
package com.mrnobody.agent.planner;

import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

/**
 * Deterministic intent routing for the unified input. A simple request should
 * never need an LLM: URLs and obvious searches are classified locally. Only
 * ambiguous or instruction-like input becomes a task.
 *
 * This is pure Java and unit-testable.
 */
public final class IntentRouter {

    // Instruction-like leading verbs: "find", "get", "download", "open … and",
    // "summarize", "compare", "check", "monitor", "every", etc.
    private static final List<String> VERBS = Arrays.asList(
            "find", "get", "fetch", "download", "summarize", "summarise",
            "compare", "check", "monitor", "track", "extract", "collect",
            "search for", "open and", "every", "watch", "scrape", "send",
            "buy", "order", "look up", "lookup",
            // "research …" / "read …" / "use google search to …" were observed
            // on-device routing to a raw results page instead of the agent.
            "research", "read", "use", "look for", "browse for");

    // Question openers: a natural-language question is an agent task, not a raw
    // search. "What is X's age" must not land on a results page.
    private static final List<String> QUESTIONS = Arrays.asList(
            "what is", "what's", "what are", "what was", "what were", "what about",
            "who is", "who's", "who are", "who was",
            "how old", "how much", "how many", "how to", "how do", "how does",
            "how is", "how are", "how long", "how tall",
            "when is", "when was", "when did", "when does", "when will",
            "where is", "where can", "where do", "where are",
            "why is", "why does", "why are", "why do",
            "which is", "which one", "which are",
            "tell me", "explain", "define", "is", "are", "does", "did", "can you");

    // Trailing fact-words: "hrithik roshan age", "bitcoin price today".
    private static final List<String> FACT_WORDS = Arrays.asList(
            "age", "height", "birthday", "birthdate", "net worth", "worth", "price",
            "salary", "girlfriend", "wife", "husband", "nationality", "religion",
            "instagram", "twitter", "meaning", "definition", "capital", "population");

    private static final Pattern IP = Pattern.compile("^\\d{1,3}(\\.\\d{1,3}){3}(:\\d+)?$");
    private static final Pattern LOCALHOST = Pattern.compile("^localhost(:\\d+)?$");
    // Any scheme, e.g. https://, http://, mailto:, intent:// — pure Java so the
    // router stays JVM-testable (no android.webkit dependency).
    private static final Pattern SCHEME = Pattern.compile("^[a-zA-Z][a-zA-Z0-9+.-]*://.*");

    private IntentRouter() {
    }

    /** Classify one line of input. Never throws. */
    public static IntentType route(String input) {
        if (input == null) return IntentType.SEARCH;
        String s = input.trim();
        if (s.isEmpty()) return IntentType.SEARCH;

        // Slash commands are explicit user overrides: heuristics will always
        // miss some phrasing, so "/agent …" must be deterministic.
        IntentType slash = slashCommand(s);
        if (slash != null) return slash;

        // explicit scheme → URL
        if (SCHEME.matcher(s).matches()) return IntentType.URL;

        // bare hostname / IP / localhost → URL
        if (IP.matcher(s).matches() || LOCALHOST.matcher(s).matches()) return IntentType.URL;
        if (looksLikeDomain(s)) return IntentType.URL;

        // instruction-like → task
        String lower = s.toLowerCase(Locale.ROOT);
        for (String v : VERBS) {
            if (lower.startsWith(v + " ") || lower.equals(v)) return IntentType.TASK;
        }
        // "open <url> and download" style compound instructions
        if (lower.contains(" and ")) return IntentType.TASK;

        // a question → task
        for (String q : QUESTIONS) {
            if (lower.equals(q) || lower.startsWith(q + " ")) return IntentType.TASK;
        }

        // a vague/partial fact lookup ("hrithik roshan age") → task
        String[] words = lower.split("\\s+");
        if (words.length >= 2 && FACT_WORDS.contains(words[words.length - 1])) {
            return IntentType.TASK;
        }

        // otherwise a search
        return IntentType.SEARCH;
    }

    private static boolean looksLikeDomain(String s) {
        if (s.contains(" ")) return false;
        // isolate the host: strip any path, then any port
        String host = s;
        int slash = host.indexOf('/');
        if (slash >= 0) host = host.substring(0, slash);
        int colon = host.indexOf(':');
        if (colon >= 0) host = host.substring(0, colon);

        int dot = host.indexOf('.');
        if (dot <= 0 || dot == host.length() - 1) return false;
        // has a dot and a plausible TLD (letters only, no spaces)
        String tld = host.substring(host.lastIndexOf('.') + 1);
        return tld.matches("[a-zA-Z]{2,}");
    }

    /**
     * Explicit slash-command routing: {@code /agent} and {@code /task} force
     * the agent, {@code /download} forces a download task, {@code /search}
     * forces a plain browser search, {@code /open} forces URL handling.
     *
     * @return the forced type, or null when input is not a slash command
     */
    public static IntentType slashCommand(String input) {
        if (input == null) return null;
        String lower = input.trim().toLowerCase(Locale.ROOT);
        if (lower.startsWith("/agent ") || lower.startsWith("/task ")
                || lower.startsWith("/download ") || lower.startsWith("/dl ")) {
            return IntentType.TASK;
        }
        if (lower.startsWith("/search ")) return IntentType.SEARCH;
        if (lower.startsWith("/open ")) return IntentType.URL;
        return null;
    }

    /**
     * The text a slash command carries: {@code /agent why is the sky blue} →
     * {@code why is the sky blue}; {@code /download <url>} →
     * {@code download <url>} so the existing download routing applies.
     * Non-command input returns unchanged (trimmed).
     */
    public static String payload(String input) {
        if (input == null) return "";
        String s = input.trim();
        String lower = s.toLowerCase(Locale.ROOT);
        for (String prefix : Arrays.asList("/agent ", "/task ", "/search ", "/open ")) {
            if (lower.startsWith(prefix)) return s.substring(prefix.length()).trim();
        }
        if (lower.startsWith("/download ")) {
            return "download " + s.substring("/download ".length()).trim();
        }
        if (lower.startsWith("/dl ")) {
            return "download " + s.substring("/dl ".length()).trim();
        }
        return s;
    }
}

File: app/lib/router/intent_router.dart
/// Deterministic intent classification for the unified input — a Dart mirror of
/// the Java IntentRouter (same rules, pure logic, unit-testable).
enum IntentType { url, search, task }

class IntentRouter {
  static const _verbs = {
    'find', 'get', 'fetch', 'download', 'summarize', 'summarise', 'compare',
    'check', 'monitor', 'track', 'extract', 'collect', 'search for',
    'open and', 'every', 'watch', 'scrape', 'send', 'buy', 'order', 'look up',
    // "research the tallest buildings" and "use google search to find …"
    // were observed routing to a raw results page instead of the agent.
    'research', 'read', 'use', 'look for', 'browse for',
  };

  /// Question openers: a natural-language question is an agent task, not a
  /// raw search. "What is X's age" went to a DuckDuckGo results page because
  /// nothing recognised it as a question; it should go to the agent, which
  /// searches and answers instead of dumping a results page.
  static const _questions = {
    'what is', "what's", 'what are', 'what was', 'what were', 'what about',
    'who is', "who's", 'who are', 'who was',
    'how old', 'how much', 'how many', 'how to', 'how do', 'how does',
    'how is', 'how are', 'how long', 'how tall',
    'when is', 'when was', 'when did', 'when does', 'when will',
    'where is', 'where can', 'where do', 'where are',
    'why is', 'why does', 'why are', 'why do',
    'which is', 'which one', 'which are',
    'tell me', 'explain', 'define', 'is', 'are', 'does', 'did', 'can you',
  };

  /// Trailing fact-words: a phrase ending in one of these is a lookup the
  /// agent should handle even without a question word ("hrithik roshan age").
  static const _factWords = {
    'age', 'height', 'birthday', 'birthdate', 'net worth', 'worth', 'price',
    'salary', 'girlfriend', 'wife', 'husband', 'nationality', 'religion',
    'instagram', 'twitter', 'meaning', 'definition', 'capital', 'population',
  };

  static final _scheme = RegExp(r'^[a-zA-Z][a-zA-Z0-9+.-]*://');
  static final _ip = RegExp(r'^\d{1,3}(\.\d{1,3}){3}(:\d+)?$');
  static final _localhost = RegExp(r'^localhost(:\d+)?$');

  static IntentType route(String input) {
    final s = input.trim();
    if (s.isEmpty) return IntentType.search;
    // Slash commands are explicit user overrides: classification heuristics
    // will always miss some phrasing, so "/agent …" must be deterministic.
    final slash = slashCommand(s);
    if (slash != null) return slash;
    if (_scheme.hasMatch(s)) return IntentType.url;
    if (_ip.hasMatch(s) || _localhost.hasMatch(s)) return IntentType.url;
    if (_looksLikeDomain(s)) return IntentType.url;

    final lower = s.toLowerCase();
    for (final v in _verbs) {
      if (lower == v || lower.startsWith('$v ')) return IntentType.task;
    }
    if (lower.contains(' and ')) return IntentType.task;
    for (final q in _questions) {
      if (lower == q || lower.startsWith('$q ')) return IntentType.task;
    }
    // A vague/partial lookup: "hrithik roshan age", "bitcoin price today".
    final words = lower.split(' ');
    if (words.length >= 2 && _factWords.contains(words.last)) {
      return IntentType.task;
    }
    return IntentType.search;
  }

  /// Explicit slash-command routing: `/agent` and `/task` force the agent,
  /// `/download <x>` forces a download task, `/search` forces a plain browser
  /// search, `/open` forces URL handling. Returns null when [input] is not a
  /// slash command.
  static IntentType? slashCommand(String input) {
    final lower = input.trim().toLowerCase();
    if (lower.startsWith('/agent ') || lower.startsWith('/task ') ||
        lower.startsWith('/download ') || lower.startsWith('/dl ')) {
      return IntentType.task;
    }
    if (lower.startsWith('/search ')) return IntentType.search;
    if (lower.startsWith('/open ')) return IntentType.url;
    return null;
  }

  /// The text a slash command carries: `/agent why is the sky blue` →
  /// `why is the sky blue`; `/download <url>` → `download <url>` so the
  /// existing download routing applies. Non-command input returns unchanged.
  static String payload(String input) {
    final s = input.trim();
    final lower = s.toLowerCase();
    for (final prefix in const ['/agent ', '/task ', '/search ', '/open ']) {
      if (lower.startsWith(prefix)) return s.substring(prefix.length).trim();
    }
    if (lower.startsWith('/download ')) {
      return 'download ${s.substring('/download '.length).trim()}';
    }
    if (lower.startsWith('/dl ')) {
      return 'download ${s.substring('/dl '.length).trim()}';
    }
    return s;
  }

  /// Normalize input into a full URL: bare domains get https://, else it's a
  /// search query for [searchEngine] — the user's configured engine, with
  /// DuckDuckGo only as the fallback when none is supplied.
  static String toUrl(String input, {String? searchEngine}) {
    final s = input.trim();
    if (s.toLowerCase().startsWith('http://')) {
      return 'https://${s.substring('http://'.length)}';
    }
    if (_scheme.hasMatch(s)) return s;
    if (_ip.hasMatch(s) || _localhost.hasMatch(s)) return 'https://$s';
    if (_looksLikeDomain(s)) return 'https://$s';
    return _searchUrl(searchEngine, s);
  }

  /// Build a search URL from the configured engine. The app's engine values
  /// are URL prefixes ending in `?q=` (AppState.searchEngines, mirrored by
  /// Settings.SEARCH_* in the core); `{q}` templates and bare hosts get
  /// handled too so a future engine shape cannot silently mis-route.
  static String _searchUrl(String? engine, String query) {
    final e = (engine == null || engine.trim().isEmpty)
        ? 'https://duckduckgo.com/?q='
        : engine.trim();
    final q = Uri.encodeComponent(query);
    if (e.contains('{q}')) return e.replaceFirst('{q}', q);
    if (e.endsWith('=')) return '$e$q';
    if (e.endsWith('?')) return '${e}q=$q';
    return '${e.contains('?') ? '$e&q=' : '$e?q='}$q';
  }

  static bool _looksLikeDomain(String s) {
    if (s.contains(' ')) return false;
    final slash = s.indexOf('/');
    var host = slash >= 0 ? s.substring(0, slash) : s;
    final colon = host.indexOf(':');
    if (colon >= 0) host = host.substring(0, colon);
    final dot = host.indexOf('.');
    if (dot <= 0 || dot == host.length - 1) return false;
    return RegExp(r'^[a-zA-Z]{2,}$').hasMatch(host.substring(host.lastIndexOf('.') + 1));
  }
}
